internal/faultinjectors/mirroring_test.go

```diff
@@ -144,6 +148,172 @@ func TestMirroring(t *testing.T) {
 	})
 }
 
+func TestMirrorParams_Address_AlreadyAnAttachFrame(t *testing.T) {
+	td := []struct {
+		Out             bool
+		Role            encoding.Role
+		ExpectedAddress string
+	}{
+		{Out: true, Role: encoding.RoleReceiver, ExpectedAddress: "source-address"},
+		{Out: false, Role: encoding.RoleReceiver, ExpectedAddress: "target-address"},
+		{Out: true, Role: encoding.RoleSender, ExpectedAddress: "target-address"},
+		{Out: false, Role: encoding.RoleSender, ExpectedAddress: "source-address"},
+	}
+
+	for _, d := range td {
+		t.Run(fmt.Sprintf("ATTACH frame: %v", d), func(t *testing.T) {
+			body := &frames.PerformAttach{
+				Name:   "name",
+				Role:   d.Role,
+				Source: &frames.Source{Address: "source-address"},
+				Target: &frames.Target{Address: "target-address"},
+			}
+
+			p := MirrorCallbackParams{
+				Out:   d.Out,
+				Frame: &frames.Frame{Body: body},
+			}
+
+			require.Equal(t, d.ExpectedAddress, p.Address())
+			require.NotNil(t, p.attachFrame)
+			require.Equal(t, d.ExpectedAddress, p.Address())
+		})
+	}
+}
+
+func TestMirrorParams_Address_LinkFrame(t *testing.T) {
+	t.Run("sender attach", func(t *testing.T) {
+		sm := loadStateMap(t)
+
+		p := MirrorCallbackParams{
+			Out: true,
+			Frame: &frames.Frame{
+				Header: frames.Header{Channel: 300},
+				Body: &frames.PerformFlow{
+					Handle: utils.Ptr(uint32(300)),
+				},
+			},
+			StateMap: sm,
+		}
+
+		require.Equal(t, "testQueue", p.Address())
+
+		p = MirrorCallbackParams{
+			Out: false,
+			Frame: &frames.Frame{
+				Header: frames.Header{Channel: 1001},
+				Body: &frames.PerformFlow{
+					Handle: utils.Ptr(uint32(1002)),
+				},
+			},
+			StateMap: sm,
+		}
+
+		require.Equal(t, "testQueue", p.Address())
+	})
+
+	t.Run("receiver attach", func(t *testing.T) {
+		sm := loadStateMap(t)
+
+		p := MirrorCallbackParams{
+			Out: true,
+			Frame: &frames.Frame{
+				Header: frames.Header{Channel: 200},
+				Body: &frames.PerformFlow{
+					Handle: utils.Ptr(uint32(200)),
+				},
+			},
+			StateMap: sm,
+		}
+
+		require.Equal(t, "testQueue", p.Address())
+
+		p = MirrorCallbackParams{
+			Out: false,
+			Frame: &frames.Frame{
+				Header: frames.Header{Channel: 0},
+				Body: &frames.PerformFlow{
+					Handle: utils.Ptr(uint32(0)),
+				},
+			},
+			StateMap: sm,
+		}
+
+		require.Equal(t, "testQueue", p.Address())
+	})
+
+	// localChannel, localHandle := uint16(150), uint32(100)
```

Comment:
are these intentionally commented out?

Reply:
Of course!

(pushes additional commit)